// Code generated by gotemplate. DO NOT EDIT.

package exception

import (
	"bytes"
	"encoding/json"
	"reflect"
	"strconv"

	"github.com/eosspark/eos-go/log"
)

// template type Exception(PARENT,CODE,WHAT)

var OutOfRangeExceptionName = reflect.TypeOf(OutOfRangeException{}).Name()

type OutOfRangeException struct {
	Exception
	Elog log.Messages
}

func NewOutOfRangeException(parent Exception, message log.Message) *OutOfRangeException {
	return &OutOfRangeException{parent, log.Messages{message}}
}

func (e OutOfRangeException) Code() int64 {
	return OutOfRangeExceptionCode
}

func (e OutOfRangeException) Name() string {
	return OutOfRangeExceptionName
}

func (e OutOfRangeException) What() string {
	return "Out of Range"
}

func (e *OutOfRangeException) AppendLog(l log.Message) {
	e.Elog = append(e.Elog, l)
}

func (e OutOfRangeException) GetLog() log.Messages {
	return e.Elog
}

func (e OutOfRangeException) TopMessage() string {
	for _, l := range e.Elog {
		if msg := l.GetMessage(); msg != "" {
			return msg
		}
	}
	return e.String()
}

func (e OutOfRangeException) DetailMessage() string {
	var buffer bytes.Buffer
	buffer.WriteString(strconv.Itoa(int(e.Code())))
	buffer.WriteString(" ")
	buffer.WriteString(e.Name())
	buffer.WriteString(": ")
	buffer.WriteString(e.What())
	buffer.WriteString("\n")
	for _, l := range e.Elog {
		buffer.WriteString("[")
		buffer.WriteString(l.GetMessage())
		buffer.WriteString("] ")
		buffer.WriteString(l.GetContext().String())
		buffer.WriteString("\n")
	}
	return buffer.String()
}

func (e OutOfRangeException) String() string {
	return e.DetailMessage()
}

func (e OutOfRangeException) MarshalJSON() ([]byte, error) {
	type Exception struct {
		Code int64  `json:"code"`
		Name string `json:"name"`
		What string `json:"what"`
	}

	except := Exception{
		Code: OutOfRangeExceptionCode,
		Name: OutOfRangeExceptionName,
		What: "Out of Range",
	}

	return json.Marshal(except)
}

func (e OutOfRangeException) Callback(f interface{}) bool {
	switch callback := f.(type) {
	case func(*OutOfRangeException):
		callback(&e)
		return true
	case func(OutOfRangeException):
		callback(e)
		return true
	default:
		return false
	}
}
